content/techniques/logger.md

```diff
@@ -133,30 +133,68 @@ await app.listen(3000);
 
 Here we use the `get()` method on the `NestApplication` instance to retrieve the singleton instance of the `MyLogger` object. This technique is essentially a way to "inject" an instance of a logger for use by Nest. The `app.get()` call retrieves the singleton instance of `MyLogger`, and depends on that instance being first injected in another module, as described above.
 
-You can also inject this `MyLogger` provider in your feature classes, thus ensuring consistent logging behavior across both Nest system logging and application logging. See <a href="techniques/logger#using-the-logger-for-application-logging">Using the logger for application logging</a> below for more information.
+The only downside of this solution is that your first initialization messages won't be printed anywhere at all, so in rare cases you may miss some important initialization errors.
+Alternatively you can print first initialization messages using default logger, and then switch to your custom one:
 
-The only downside of this solution is that your first initialization messages won't be handled by your logger instance, though, it shouldn't really matter at this point.
+```typescript
+const app = await NestFactory.create(ApplicationModule);
+app.useLogger(app.get(MyLogger));
+await app.listen(3000);
+``` 
+
+You can also inject this `MyLogger` provider in your feature classes, thus ensuring consistent logging behavior across both Nest system logging and application logging. See <a href="techniques/logger#using-the-logger-for-application-logging">Using the logger for application logging</a> and <a href="techniques/logger#injecting-a-custom-logger">Injecting a custom logger</a> below for more information.
 
 #### Using the logger for application logging
 
-We can combine several of the techniques above to provide consistent behavior and formatting across both Nest system logging and our own application event/message logging. In this section, we'll achieve this with the following steps:
+We can combine several of the techniques above to provide consistent behavior and formatting across both Nest system logging and our own application event/message logging.
+
+The best practice is to instantiate `Logger` class from `@nestjs/common` in each of our services. 
+We can supply our service name as the `context` argument in the `Logger` constructor, like so:
 
-1. We extend the built-in logger and customize the `context` portion of the log message (e.g., the phrase `NestFactory` in square brackets in the log line shown below).
+```typescript
+import { Logger, Injectable } from '@nestjs/common';
+
+@Injectable()
+class MyService {
+  logger = new Logger(MyService.name);
```

Comment:
It would be better to show this in the `constructor` to better follow dependency injection principles.

Reply:
Do you mean like this?
```
constructor() {
  this.logger = new Logger(MyService.name);
}
```

From my standpoint it is no different from what wrote, but through the times of my googling I only saw variant with property initializer (even in the nestjs source), so I used it here as well.

Reply:
Rather something like
```ts
class MyService {
  constructor(private readonly logger: Logger) {
    this.logger.setContext(MyService.name);
  }
}
```

Reply:
I had some issues with that with `nestjs-pino` because it has its own provider for `Logger`, so this code actually injects `Logger` from `nestjs-pino` instead of `Logger` from `@nestjs/common`. And the former does not have `setContext` function. So it produces a runtime error. Maybe this is a mistake with the library, but still `new Logger` seems more safe.

Also, `setContext` mutates `Logger` instance and `Logger` seems to have singleton scope (https://github.com/nestjs/nest/blob/master/packages/common/services/logger.service.ts#L19), so I am not sure this would work.

Reply:
Ah, in the case of singleton scopes, you'd be right about that. In the case of `nestjs-pino` I feel that's on the library. You could always get around that with `@Inject()`, but that's not what this PR is about ;). All right, LGTM.